text/0000-macro-hygiene-optout.md

```diff
@@ -0,0 +1,229 @@
+- Feature Name: macro_hygiene_optout
+- Start Date: 2018-07-05
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This feature introduces the ability to "opt-out" of the usual macro hygiene rules within definitions of [declarative macros][rfc-decl-macro] (macros 2.0), for designated occurrences of identifiers. In other words, the feature will enable one to annotate occurrences of identifiers with macro call-site hygiene rather than the default definition-site hygiene.
+
+# Motivation
+[motivation]: #motivation
+
+The use of [hygienic macros][rust-hygienic-macros] in Rust is justified by much prior research and experience <sup>([1][paper-hygienic-macro-expansion] [2][scheme-hygiene])</sup>, and solves several common issues that programmers would otherwise encounter with macros due to the nature of syntactical substitution. The principal deficit of this approach is that it requires that names/identifiers of any items generated by a macro be *explicitly passed to* the macro as arguments. This both requires the logic for name selection to remain entirely external to the macro, and even if that is not a problem, the passing of all identifiers-to-export into a macro can quickly become unwieldy for macros that generate many identifiers.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Escaping of hygiene for identifiers within macros allows one to define identifiers with syntax contexts (**hygiene**) corresponding to the location in the source code from which the macro is invoked (the **call-site**) rather than the location it is defined (**definition-site**). It also enables one to use/reference existing identifiers from the call-site from within macro definitions, though this is not the true aim of the feature, but rather a side-effect, and will be discussed later.
+
+Note that for the purposes of this RFC, an **identifier** can roughly be considered to be an textual name (e.g. `foo_bar`) of any sort (for a variable, function, trait, etc.) or a lifetime (e.g. `'a`).
```

Comment:
Currently all lifetime parameters are unhygienic, not sure if we will fix that for macros 2.0 or not.

Reply:
Yeah. Hopefully we will!

Reply:
Lifetimes are already hygienic in `macro` macros and with `Span::def_site()` in proc macros.